### DISCUSSION TITLE
How do I get the logs for a service?

### DISCUSSION THREAD

**Discussion by colinhowetab** [community/community#26282] [Actions]
<p>Hi,</p>
<p>I’ve got several services that my job depends on. I’m fairly certain one of them is misbehaving. I’d like to view the logs for them, but, I can’t see them anywhere within github actions. Any suggestions?</p>
<p>Cheers!</p>

**Comment by paramt** [comment]
<p>At the top right corner of the Actions tab, there is a link to view the raw logs. </p>
<p><img src="https://user-images.githubusercontent.com/23563074/181096641-65426958-48a2-49fd-9ceb-aff988b95972.png" alt="image.png" data-base62-sha1="azMcPeZRJshpD8d2BMChPZs53LF" width="394" height="178"></p>

**Comment by dakale** [comment, marked as answer]
<p><a href="https://help.github.com/en/articles/contexts-and-expression-syntax-for-github-actions#job-context" rel="nofollow noopener">https://help.github.com/en/articles/contexts-and-expression-syntax-for-github-actions#job-context</a></p>
<p>You can get the ID of a service container with “${{ job.services.&lt;service_id&gt;.id }}”</p>
<p>With that you can do any regular docker operation you’d like, for example:</p>
<pre><code>docker logs "${{ job.services.&lt;service_id&gt;.id }}"
</code></pre>
<p>Heres a full working sample I have of using mysql:</p>
<pre><code>jobs:
  test:
    runs-on: ubuntu-latest
    services:
      mysql:
        image: mysql:5.7
        env:
          MYSQL_ROOT_PASSWORD: password
        ports:
          - 3306
        options: --health-cmd="mysqladmin ping" --health-interval=10s --health-timeout=5s --health-retries=3
    steps:
    - name: Verify MySQL connection from host
      run: |
        echo "${{ toJson(job) }}"
        sudo apt-get install -y mysql-client
        mysql --host 127.0.0.1 --port ${{ job.services.mysql.ports['3306'] }} -uroot -ppassword -e "SHOW DATABASES"
        docker logs "${{ job.services.mysql.id }}"
</code></pre>
<p>The logs from the service container will then be in the logs of the step where you ran <code>docker logs</code>, as expected </p>
<p>Depending on the service, <code>docker logs</code> may not work. See: <a href="https://docs.docker.com/v17.09/engine/admin/logging/view_container_logs/" rel="nofollow noopener">https://docs.docker.com/v17.09/engine/admin/logging/view_container_logs/</a></p>

**Comment by colinhowetab** [comment]
<p>Thank you!</p>
<p>Not sure where to put feature requests… would be good to have these automatically extracted and viewable within a tab somewhere in the UI. When a build fails I’d rather not have to start adding calls to docker to try and diagnose the failure :slight_smile:</p>

**Comment by teohhanhui** [comment]
<p>This would not work if the “Initialize containers” step has failed. How do we troubleshoot such a failure?</p>
<p>See for example <a href="https://github.com/api-platform/core/runs/260470103#step:2:133" rel="nofollow noopener">https://github.com/api-platform/core/runs/260470103#step:2:133</a></p>

**Comment by jmalloc** [comment]
<p>I’d like to know this too, I am having a similar issue with a MariaDB service (which incidentally was working fine a couple of days ago).</p>

**Comment by kurkop** [comment]
<p>I’d like to have way to see logs for all these services</p>

**Comment by germa89** [comment]
<p>Hi</p>
<p>We solved this using <code>jwalton/gh-docker-logs@v1</code> action.</p>
<pre><code class="lang-yaml">
  - name: Collect docker logs on failure
    if: "!success()"
    uses: jwalton/gh-docker-logs@v1
    with:
      dest: './logs-docs'

  - name: Collecting documentation logs
    if: "!success()"
    continue-on-error: true
    run: |
      cp ./doc/build/latex/*.log ./logs-docs
      cp ./doc/build/latex/*.out ./logs-docs

  - name: Tar logs
    if: "!success()"
    run: tar cvzf ./logs-docs.tgz ./logs-docs

  - name: Upload logs to GitHub
    if: "!success()"
    uses: actions/upload-artifact@master
    with:
      name: logs-docs.tgz
      path: ./logs-docs.tgz
</code></pre>
